File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/Interpreter/TerminalExpression.go
// 终结符表达式 - 处理数字
package Interpreter

type NumberExpr struct {
	num float64
}

func NewNumberExpr(num float64) *NumberExpr {
	return &NumberExpr{num: num}
}

func (n *NumberExpr) Interpret() float64 {
	return n.num
}

File: DesignPatterns/01_GOF23/DP_examples/C/03_Behavioral/Interpreter/TerminalExpression.c
// 终结符表达式实现
#include "TerminalExpression.h"
#include <stdlib.h>

// 解释方法
double number_expr_interpret(ArithmeticExpr* self) {
    NumberExpr* expr = (NumberExpr*)self;
    return expr->num;
}

// 销毁方法
void number_expr_destroy(ArithmeticExpr* self) {
    free(self);
}

// 构造函数
NumberExpr* number_expr_create(double num) {
    NumberExpr* expr = (NumberExpr*)malloc(sizeof(NumberExpr));
    if (expr) {
        expr->base.interpret = number_expr_interpret;
        expr->base.destroy = number_expr_destroy;
        expr->num = num;
    }
    return expr;
}